src/ssl.c

```diff
@@ -51319,6 +51342,608 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
+        MIME_free_hdrs(allHdrs);
+        if (section)
+            XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+        wolfSSL_BIO_free(*bcont);
+        return NULL;
+    }
+    outLen = ((sectionLen*3+3)/4)+1;
+    out = (byte*)XMALLOC(outLen*sizeof(byte), NULL, DYNAMIC_TYPE_PKCS7);
+    outHead = out;
+    if (!outHead) {
+        MIME_free_hdrs(allHdrs);
+        if (outHead)
+            XFREE(outHead, NULL, DYNAMIC_TYPE_PKCS7);
+        if (section)
+            XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+        wolfSSL_BIO_free(*bcont);
+        return NULL;
+    }
+    /* Strip trailing newlines. */
+    while ((section[sectionLen-1] == '\r' || section[sectionLen-1] == '\n') &&
+           sectionLen > 0) {
+        sectionLen--;
+    }
+    section[sectionLen] = '\0';
+    ret = Base64_Decode((const byte*)section, sectionLen, out, &outLen);
+    if (ret < 0) {
+        WOLFSSL_MSG("Error base64 decoding S/MIME message.\n");
+        MIME_free_hdrs(allHdrs);
+        XFREE(outHead, NULL, DYNAMIC_TYPE_PKCS7);
+        XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+        wolfSSL_BIO_free(*bcont);
+        return NULL;
+    }
+    pkcs7 = wolfSSL_d2i_PKCS7_ex(NULL, (const unsigned char**)&out, outLen,
+                                 bcontMem, bcontMemSz);
+
+
+    MIME_free_hdrs(allHdrs);
+    XFREE(outHead, NULL, DYNAMIC_TYPE_PKCS7);
+    XFREE(section, NULL, DYNAMIC_TYPE_PKCS7);
+
+
+    return pkcs7;
+}
+
+/*****************************************************************************
+* MIME_parse_headers - Reads the BIO in and parses out MIME headers and
+* parameters into headers.  Will continue until in has no more content, inLen
+* characters are processed or a double CR/LF is reached.
+*
+* RETURNS:
+* returns zero on success, non-zero on error.
+*/
+int MIME_parse_headers(WOLFSSL_BIO* in, int inLen, MimeHdr** headers)
+{
+    MimeHdr* nextHdr = NULL;
+    MimeHdr* curHdr = NULL;
+    MimeParam* nextParam = NULL;
+    word32 start = 0;
+    word32 end = 0;
+    char* nameAttr = NULL;
+    char* bodyVal = NULL;
+    MimeTypes mimeType = MIME_HDR;
+    MimeStatus mimeStatus = MIME_NAMEATTR;
+    int ret = 0;
+    int pos = 0;
+    int lineLen = 0;
+    byte isEnd = 0;
+    char* lines = NULL;
+    char* curLine = NULL;
+
+    if (!in || inLen <= 0) {
+        return ASN_PARSE_E;
+    }
+    nextHdr = (MimeHdr*)XMALLOC(sizeof(MimeHdr), NULL, DYNAMIC_TYPE_PKCS7);
+    nextParam = (MimeParam*)XMALLOC(sizeof(MimeParam), NULL,
+                                    DYNAMIC_TYPE_PKCS7);
+    lines = (char*)XMALLOC((inLen+1)*sizeof(char), NULL, DYNAMIC_TYPE_PKCS7);
+    curLine = lines;
+    if (!nextHdr || !nextParam || !lines) {
+        if (lines)
+            XFREE(lines, NULL, DYNAMIC_TYPE_PKCS7);
+        if (nextHdr)
+            XFREE(nextHdr, NULL, DYNAMIC_TYPE_PKCS7);
+        if (nextParam)
+            XFREE(nextParam, NULL, DYNAMIC_TYPE_PKCS7);
+        return MEMORY_E;
+    }
+    ForceZero(nextHdr, (word32)sizeof(MimeHdr));
```

Comment:
Do not use ForceZero unless you are clearing sensitive data. Instead use `XMEMSET`. This will let the compiler optimize better.